python/xir/program.py

```diff
@@ -567,6 +567,65 @@ def serialize(self, minimize: bool = False) -> str:
             return strip(res_script)
         return res_script
 
+    @staticmethod
+    def merge(*programs) -> "XIRProgram":
+        """Merges one or more XIR programs into a new XIR program.
+
+        The merged XIR program is formed by concatenating the given XIR programs
+        in the order they are passed to the function. Warnings may be issued for
+        duplicate declarations, gates, includes, operators, and options.
+
+        Args:
+            programs (XIRProgram): XIR programs to merge
+
+        Returns:
+            XIRProgram: the merged XIR program
+
+        Raises:
+            ValueError: if no XIR programs are provided or if at least two XIR
+                programs have different versions or float settings
+        """
+        if len(programs) == 0:
+            raise ValueError("Merging requires at least one XIR program.")
+
+        version = programs[0].version
+        if any(program.version != version for program in programs):
+            raise ValueError("XIR programs with different versions cannot be merged.")
+
+        use_floats = programs[0].use_floats
+        if any(program.use_floats != use_floats for program in programs):
+            raise ValueError("XIR programs with different float settings cannot be merged.")
```

Comment:
For my understanding - what is `use_floats` and the implications of it being `True` or `False`?

Reply:
Good question! Basically, `use_floats` determines whether floating-point parameters in an XIR program are returned as `float` (`complex`) objects or `Decimal` (`DecimalComplex`) objects.  The former is more convenient while the latter is more precise.

Reply:
> So this only affects how floats/complex are returned, and could potentially be switched on or off (although currently there's no setter method for this).

True although we may as well just make it a public attribute at that point.

> An option would be to just automatically set it to `True` or `False` as default, and provide a warning if it's `False` in either of the merged programs. This would mean that there's less of an annoyance if a user has multiple programs that happen to use different settings here, since there's no real reason not to allow merging other than this rather flexible boolean.

Not a bad idea! Perhaps we should set `use_floats` to the OR of the merged XIR programs in order to preserve its value when it is shared by all merged XIR programs; OR is preferred to AND since losing precision is viral.

> Although this would mean that all statements also need to be added with the same `use_floats`. Actually, maybe the `add_statement` function should validate that all added statements have the same value for `use_floats` as the program they're added to, in that case that wouldn't be an issue. What do you think?

Sure! It will probably be simpler to just override the value of `use_floats` in the `Statement` passed to `XIRProgram.add_statement()`. An alternative approach is to have the validator ensure that `Statement.use_floats` and `OperatorStmt.use_floats` are consistent with `XIRProgram.use_floats`.

Reply:
**Update:** I've modified `XIRProgram.merge()` to take the OR of the float settings and downgraded the error into a warning.

Reply:
> It will probably be simpler to just override the value of use_floats in the Statement passed to XIRProgram.add_statement()

I would probably go for this!